backend/src/pong/pong.service.ts

```diff
@@ -0,0 +1,156 @@
+import { Injectable } from '@nestjs/common';
+import { Ball, GameState, MovementKeys, Paddle, Position } from './Game.types';
+
+@Injectable()
+export class PongService {
+	private readonly paddleWidth = 1;
+	private readonly paddleHeight = 12;
+
+	private readonly ballRadius = 1;
+
+	// ball direction and speed %
+	private dx = -0.4;
+	private dy = 0.4;
+
+	createNewGameState(): GameState {
+		const newGameState: GameState = {
+			playerOnePaddle: {
+				position: {
+					x: 0,
+					y: 50,
+				},
+				width: this.paddleWidth,
+				height: this.paddleHeight,
+			},
+
+			playerTwoPaddle: {
+				position: {
+					x: 100,
+					y: 50,
+				},
+				width: this.paddleWidth,
+				height: this.paddleHeight,
+			},
+
+			ball: {
+				position: {
+					x: 40,
+					y: 65,
+				},
+				radius: this.ballRadius,
+			},
+			canvas: {
+				x: 1200,
+				y: 960,
+			},
+		};
+		return newGameState;
+	}
+
+	doesHitWall(radius: number, pos: number, direction: number): boolean {
+		const fullHeigthOrWidth = 100;
+		return (
+			pos + direction > fullHeigthOrWidth - radius || pos + direction < radius
+		);
+	}
+
+	doesHitPaddle(ball: Ball, rect: Paddle, canvas: Position) {
+		const paddleY = rect.position.y;
+		const paddleHeight = rect.height;
+		const paddleTop = paddleY - paddleHeight / 2;
+		const paddleBottom = paddleY + paddleHeight / 2;
+		const ballLineThicknessOffset = 0.2;
+		const paddleWall =
+			ball.radius - ballLineThicknessOffset + this.paddleWidth / 2;
+
+		// if ball will hit the invisible wall that extends from the paddle
+		// and ball position is within the paddle top and bottom
+		// console.log(paddleWall);
+		if (
+			this.doesHitWall(paddleWall, ball.position.x, this.dx) &&
```

Comment:
I do have the feeling that the balls goes a little bit through the walls top and bottom for the rest it looks nicE!

Reply:
cool